consensus/cryptoiface.go

```diff
@@ -0,0 +1,82 @@
+package consensus
+
+import (
+	"github.com/relab/hotstuff"
+)
+
+// CryptoBase provides the basic cryptographic methods needed to create, verify, and combine signatures.
+type CryptoBase interface {
+	// Sign creates a cryptographic signature of the given message.
+	Sign(message []byte) (signature QuorumSignature, err error)
+	// Verify verifies the given cryptographic signature according to the specified options.
+	// NOTE: One of either VerifySingle or VerifyMulti options MUST be specified,
+	// otherwise this function will have nothing to verify the signature against.
+	Verify(signature QuorumSignature, options ...VerifyOption) bool
+	// Combine combines multiple signatures into a single signature.
+	Combine(signatures ...QuorumSignature) (signature QuorumSignature, err error)
+}
+
+// VerifyOption sets options for the Verify function in the CryptoBase interface.
+type VerifyOption func(o *VerifyOptions)
+
+// VerifyThreshold specifies the number of signers that must have participated
+// in the signing of the signature in order for Verify to return true.
+func VerifyThreshold(threshold int) VerifyOption {
+	return func(o *VerifyOptions) {
+		o.Threshold = threshold
+	}
+}
+
+// VerifySingle verifies the signature against the given message.
+//
+// NOTE: VerifySingle and VerifyMulti cannot be used together,
+// and whichever option is specified last will be the effective one.
+func VerifySingle(message []byte) VerifyOption {
+	return func(o *VerifyOptions) {
+		o.MultipleMessages = false
+		o.Message = &message
+	}
+}
+
+// VerifyMulti verifies the signature against the multiple messages present in the given map.
+//
+// NOTE: VerifySingle and VerifyMulti cannot be used together,
+// and whichever option is specified last will be the effective one.
+func VerifyMulti(messages map[hotstuff.ID][]byte) VerifyOption {
+	return func(o *VerifyOptions) {
+		o.MultipleMessages = true
+		o.MessageMap = messages
+	}
+}
+
+// VerifyOptions specify options for the Verify function in the CryptoBase interface.
+type VerifyOptions struct {
+	// using a pointer for the Message such that we can tell if it is unspecified
```

Comment:
I think `Message` does not need to be a pointer. You can still distinguish these two cases, if needed:
```go
b = []byte{} // non-nil, but len(b) == 0
b = nil // nil and len(b) == 0
```
I tried to remove the pointer and update the relevant code and the tests in the `crypto` package still pass.